### ISSUE TITLE
carbon-components 10.9.3 introduces breaking changes to UIShell

### ISSUE THREAD

**Issue by zevioso** [carbon-design-system/carbon#5287]
## What package(s) are you using?

- [x] `carbon-components`
- [x] `carbon-components-react`

## Detailed description

> Describe in detail the issue you're having.

The latest **PATCH** version of carbon-components introduced a breaking change that is not backwards-compatible. https://github.com/carbon-design-system/carbon/commit/d9935c3500ab3a04e7472376e34563d6c25076de#diff-f8c1a41d640bb6e6c4213525998e936e

Because of the level of specificity that was used to style these elements, consumers that need to customize these elements would also need to apply the same level of specificity. 

Removing these attributes will now break any CSS that utilized them. 

> Is this issue related to a specific component?

UIShell - Header

> What did you expect to happen? What happened instead? What would you like to
> see changed?

Following SEMVER this change should have included a backward-compatible solution or should have been released under a **MAJOR** version.

> What browser are you working in?

Chrome, Safari, Firefox

> What version of the Carbon Design System are you using?

10.9.3/7.9.3

## Additional information

A public example that would be affected after the upgrade - https://github.com/carbon-design-system/ibm-dotcom-library/blob/master/packages/styles/scss/components/masthead/_masthead-l1.scss#L145

**Comment by asudoh** [comment]
Hi :wave: thank you for reporting and my apologies for what the change caused. I see the pain. The change was needed to improve the a11y. Given such reason, it’s hard for us to bring the specific attribute back, but we can add a class name to the elements where `[role=“menuitem”]` were so we have the same level of specificity. Let us know if it helps.

**Comment by joshblack** [comment]
Hi @zevioso! 👋 

So sorry that this happened in a patch release, definitely understand how annoying/frustrating it can be for breakage to occur in a non-major release from a package 😞 

This is definitely a tricky one with respect to semver as the implications would be broad if we included targeting selectors as part of the public API. For example, if we moved something from a `flex` container to a different display style then it would be considered a breaking change:

```scss
// v1
.library-component {
  display: flex;
}

.my-override .library-component {
  flex-direction: row-reverse;
}

// v2
.library-component {
  display: inline-block;
}

// This now is a noop and would break the layout
.my-override .library-component {
  flex-direction: row-reverse;
}
```

In addition, with override selectors there is no way we could test to see if a change we make could break how someone is overriding a specific selector. This would essentially freeze all aspects of our styles and would make it challenging to update without considering it a breaking change.

For the reasons mentioned above, it seems like we won't consider selectors as part of the public API. They should be treated as internals/implementation details. This is definitely not an ideal situation, as clearly there is a necessity to make these kinds of modifications.

As a result, we recommend reaching out to the team if modifications/improvements need to be made to a component. Reaching out could mean that our team implements the feature, or that we work alongside another team to ship the feature. In the case of someone still needing to override, we'll start changing how we author our selectors to use minimal specificity and not rely too much on markup in selectors even if it's included for a11y. The only caveat would be that overriding means that you take ownership for all the styles at that point as we cannot test if a style modification we make alters consuming projects.

Definitely apologize for the wall of text, I hope that it's helpful in sharing how we're thinking about things. Would love to hear any feedback that you might have!
